### PR TITLE
Allow setting custom http request from binary addon

### PR DESCRIPTION
**Pull request by rbuehlma**
With this PR, it is possible to set a custom request for a CURL HTTP call.

In my case (pvr.teleboy) this is required to do DELETE calls against the REST-API of Teleboy to delete timer and recordings.

### CODE REVIEW

**Comment by ksooo** [comment]
Shouldn't this API extension documented in the header file(s) of the function(s) taking the new parameter value? How should an add-on dev otherwise know about it?

**Comment by ksooo** [comment]
And because this is a semantic API extension, we need a micro bump of the respective API version. Only version, not min version this time as add-ons do not need to be rebuilt. 

Reason is that an add-on should be able to determine whether it can actually use the new parameter by checking Kodi's API version number before doing  the call.

**Comment by ksooo** [comment]
I suggest to merge this PR together with #12780, so we could save the version bump here.

**Comment by rbuehlma** [comment]
> Shouldn't this API extension documented in the header file(s) of the function(s) taking the new parameter value? How should an add-on dev otherwise know about it?

I agree that would be nice. Currently its not the case for all of these options. Probably the devs will look into the Kodi source anyway (at least I did) but maybe a pointer to the file in line would be helpful? See my added commit.

> Shouldn't this API extension documented in the header file(s) of the function(s) taking the new parameter value? How should an add-on dev otherwise know about it?

Could we merge this PR (and also https://github.com/xbmc/xbmc/pull/12779) right before https://github.com/xbmc/xbmc/pull/12780? Then we get the API bump for free. Else, I would create a conflict between the PRs.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h [hunk 1]
```diff
@@ -563,7 +563,7 @@ namespace ADDON
     * @brief Adds options to the curl file created with CURLCeate
     * @param file file pointer to the file returned by CURLCeate
     * @param type option type to set
-    * @param name name of the option
+    * @param name name of the option (see https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/CurlFile.cpp#L793)
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: sooner or later, this will link to a wrong line

Comment (rbuehlma) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: Thats true, but the dev should still be able to find the required information. :) But yes, I will remove it again.

File: xbmc/filesystem/CurlFile.cpp [hunk 2]
```diff
@@ -444,6 +444,7 @@ CCurlFile::CCurlFile()
   m_httpresponse = -1;
   m_acceptCharset = "UTF-8,*;q=0.8"; /* prefer UTF-8 if available */
   m_allowRetry = true;
+  m_customRequest = "";
```
[the remark this answers is not in the document]
Comment (rbuehlma) [reply xbmc/filesystem/CurlFile.cpp]: ok... minus 1 change... :)

File: xbmc/filesystem/CurlFile.cpp [hunk 3]
```diff
@@ -830,6 +837,10 @@ void CCurlFile::ParseAndCorrectUrl(CURL &url2)
         {
           SetRequestHeader(it->first, value);
         }
+        else if (name == "customrequest")
+        {
+          m_customRequest = value;
+        }
```
[the remark this answers is not in the document]
Comment (rbuehlma) [reply xbmc/filesystem/CurlFile.cpp]: minus another change :)

File: xbmc/filesystem/CurlFile.h [hunk 4]
```diff
@@ -182,6 +182,7 @@ namespace XFILE
       std::string m_password;
       std::string m_httpauth;
       std::string m_cipherlist;
+      std::string m_customRequest;
```
[the remark this answers is not in the document]
Comment (rbuehlma) [reply xbmc/filesystem/CurlFile.h]: reduces change to three lines.. how did I miss that?!

**Comment by ksooo** [comment]
> I agree that would be nice. Currently its not the case for all of these options. 

 Sorry, this is not an argument for not at least starting to do it better. It would be even more cool and very helpful, if you would document the other options as well. ;-)

>Could we merge this PR (and also #12779) right before #12780? Then we get the API bump for free. Else, I would create a conflict between the PRs.

 This is what I suggested, yes - at least regarding #12780. #12779 is unrelated wrt API.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h (hunk 1 above)
Comment (ksooo) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: -1 for this, document it where it belongs, please.  We never do it like this for a reason.

**Comment by rbuehlma** [comment]
> Sorry, this is not an argument for not at least starting to do it better. It would be even more cool and very helpful, if you would document the other options as well. ;-)
Ok, ok, I'll do it. Are the two locations the rights ones? Is it ok, if I do it in an additional PR to not block this and the others?

**Comment by ksooo** [comment]
I would be happy to have that in this PR, tbh. We are not in a hurry.

**Comment by ksooo** [comment]
Just spotted that #12779 is also a micro API change. ;-) So yes, we should merge #12779, #12780 and #12790 all at once. I'm just waiting for the documentation improvements and then we are ready to go, imo.

**Comment by rbuehlma** [comment]
Here we are (the documentation). 

I spotted this line:
https://github.com/xbmc/xbmc/blob/e7dc7f24fc7089365e2ec77a90e62f362d0ed4da/xbmc/filesystem/CurlFile.cpp#L646
In my opinion it should read if (NOT m_skipshout)
but not sure about it and definitely not part of this PR.

**Comment by ksooo** [comment]
Thanks for the documentation update. Much appreciated. :-)

**Comment by ksooo** [comment]
No idea about the shoutcast thing, but yeah, definitely not related to this PR. You could open a separate PR for this and ask for opinions, if you like.

**Review by ksooo — APPROVED** [review]

**Comment by rbuehlma** [comment]
Thanks for your time and feedback :)

**Comment by ksooo** [comment]
Anytime. You're welcome. :-)

**Comment by ksooo** [comment]
Thanks for your very valuable contributions.

**Pull request merged**